support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -54,6 +55,53 @@ object GiftRecipient {
     ))
 }
 
+case class GiftCode private(value: String) extends AnyVal
+
+object GiftCode {
+
+  def apply(value: String): Option[GiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftPurchase {
+  def asDigiSub: Option[DigitalSubGiftPurchase] = this match { case a: DigitalSubGiftPurchase => Some(a); case _ => None }
+  def giftRecipient: GiftRecipient
+}
+object GiftPurchase {
+
+  case class WeeklyGiftPurchase(giftRecipient: WeeklyGiftRecipient) extends GiftPurchase
+  case class DigitalSubGiftPurchase(
+    giftRecipient: DigitalSubGiftRecipient,
+    giftCode: GiftCode
+  ) extends GiftPurchase
```

Comment:
new ADT to represent a product specific purchased gift, for DS only, so far it also includes the gift code.
In the next PR it will also get the `last_redemption_date` field added.